In [65]:
import pandas as pd
import numpy as np
from datetime import datetime

In [66]:
pd.options.display.max_columns = 999

In [67]:
df_final = pd.read_csv('arquivos/dadosalunos-21-08-2023.csv')

In [68]:
df_estados = pd.read_csv('arquivos/estados.csv', usecols=['codigo_uf', 'uf'])
df_municipios = pd.read_csv('arquivos/municipios.csv', usecols=['nome', 'codigo_uf'])

df_municipios['nome'] = df_municipios['nome'].str.upper().str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')

df_estados_municipios = df_municipios.merge(df_estados, on='codigo_uf', how='left')

In [69]:
colunas_loc_nasc_aluno = ['CD_ORIGEM_ALUNO', 'LT_NACIONALIDADE', 'NM_PAIS_NASC', 'SG_UF_NASC_ALUNO', 'NM_CIDADE_NASC', 'SG_UF_FERIAS', 'NM_CIDADE_FERIAS']

In [70]:
colunas_a_remover = []

In [213]:
df_final[(df_final['DT_NASC_ALUNO'] == '2003-06-16') & (df_final['CD_CURSO'] == 26)]

ANO_OCORRENCIA     CD_ALUNO  CD_CURSO  CD_ORIGEM_ALUNO CD_TURNO  \
66372            2021  10129599724        26                0        I   

      DE_NATURALIDADE DH_MATRICULA DH_SITUACAO  DT_ANO_INGRES DT_NASC_ALUNO  \
66372             NaN   2021-06-16  2022-04-25           2021    2003-06-16   

       LT_ESTADO_CIVIL LT_MODALIDADE LT_NACIONALIDADE  LT_RACA NM_CIDADE_AULA  \
66372              1.0             P                B        2        MARINGA   

      NM_CIDADE_EST2GRAU NM_CIDADE_FERIAS NM_CIDADE_NASC NM_PAIS_NASC  \
66372            MARINGA          MARINGA        MARINGA       BRASIL   

      NU_SEMESTRE2_VEST SG_SEXO_ALUNO SG_UF_AULA SG_UF_EST2GRAU SG_UF_FERIAS  \
66372                 A             M         PR             PR           PR   

      SG_UF_NASC_ALUNO ST_ALUNO TP_DEFIC_INEP TP_ESCOLA  TP_INGRES_CURSO  \
66372               PR      NaN      NNNNNNNN         P                1   

      TP_OPCAO_COTAS  F_CD_ORIGEM_ALUNO F_DT_NASC_ALUNO F_NM_PAIS_NASC  \
66372              N                  1            2003         BRASIL   

      F_SG_UF_FERIAS F_SG_UF_NASC_ALUNO  BOOL_TRANCADO  BOOL_ST_ALUNO  \
66372             PR                 PR              0            0.0   

      MULTI_ST_ALUNO F_TP_ESCOLA F_TP_OPCAO_COTAS  BOOL_COTAS  ANO_MATRICULA  \
66372    MATRICULADO           P                N           0           2021   

       ANO_NASCIMENTO  IDADE_MATRICULA  
66372            2003             18.0

In [212]:
df_final[(df_final['DT_NASC_ALUNO'] == '2004-04-12') & (df_final['CD_CURSO'] == 26)]

ANO_OCORRENCIA     CD_ALUNO  CD_CURSO  CD_ORIGEM_ALUNO CD_TURNO  \
70071               0  13159171903        26                0        I   

      DE_NATURALIDADE DH_MATRICULA DH_SITUACAO  DT_ANO_INGRES DT_NASC_ALUNO  \
70071             NaN   2022-05-16  2022-05-16           2022    2004-04-12   

       LT_ESTADO_CIVIL LT_MODALIDADE LT_NACIONALIDADE  LT_RACA NM_CIDADE_AULA  \
70071              1.0             P                B        1        MARINGA   

      NM_CIDADE_EST2GRAU NM_CIDADE_FERIAS NM_CIDADE_NASC NM_PAIS_NASC  \
70071            MARINGA          MARINGA        MARINGA       BRASIL   

      NU_SEMESTRE2_VEST SG_SEXO_ALUNO SG_UF_AULA SG_UF_EST2GRAU SG_UF_FERIAS  \
70071                 A             F         PR             PR           PR   

      SG_UF_NASC_ALUNO ST_ALUNO  TP_DEFIC_INEP TP_ESCOLA  TP_INGRES_CURSO  \
70071               PR      NaN  NNNNNNNNNNNNN         P               14   

      TP_OPCAO_COTAS  F_CD_ORIGEM_ALUNO F_DT_NASC_ALUNO F_NM_PAIS_NASC  \
70071              N                  1            2004         BRASIL   

      F_SG_UF_FERIAS F_SG_UF_NASC_ALUNO  BOOL_TRANCADO  BOOL_ST_ALUNO  \
70071             PR                 PR              0            0.0   

      MULTI_ST_ALUNO F_TP_ESCOLA F_TP_OPCAO_COTAS  BOOL_COTAS  ANO_MATRICULA  \
70071    MATRICULADO           P                N           0           2022   

       ANO_NASCIMENTO  IDADE_MATRICULA  
70071            2004             18.0

### ANO_OCORRENCIA

Não mexi e não iremos utilizar, uma vez que não há o campo da Ocorrência

In [72]:
colunas_a_remover.append('ANO_OCORRENCIA')

### CD_ALUNO

ID do aluno. Mantém como está

### CD_CURSO

Mantém como está.

### CD_ORIGEM_ALUNO

In [73]:
df_final['CD_ORIGEM_ALUNO'].value_counts(dropna=False).sort_index()

CD_ORIGEM_ALUNO
0      7322
1     27404
2      6925
3     17061
4      9967
5      2394
6       147
15        1
Name: count, dtype: int64

Os valores 0 são nulos, verificando outras colunas referentes ao local de nascimento do aluno

In [74]:
def verifica_origem_aluno(row):
    if row['NM_CIDADE_NASC'] == 'MARINGA':
        return 1
    elif row['SG_UF_NASC_ALUNO'] == 'SP':
        return 4
    elif row['SG_UF_NASC_ALUNO'] != 'SP' and row['SG_UF_NASC_ALUNO'] != 'PR':
        return 5
    elif row['NM_PAIS_NASC'] != 'BRASIL':
        return 6
    else:
        return 0

In [75]:
df_final['F_CD_ORIGEM_ALUNO'] = df_final.apply(verifica_origem_aluno, axis=1)

In [76]:
df_final[df_final['CD_ORIGEM_ALUNO'] == 0][colunas_loc_nasc_aluno + ['F_CD_ORIGEM_ALUNO']]

CD_ORIGEM_ALUNO LT_NACIONALIDADE NM_PAIS_NASC SG_UF_NASC_ALUNO  \
1                    0                B       BRASIL               PR   
51                   0                B       BRASIL               PR   
809                  0                B       BRASIL               PR   
1489                 0                B       BRASIL               PR   
1490                 0                B       BRASIL               PR   
...                ...              ...          ...              ...   
71216                0                B       BRASIL               PR   
71217                0                B       BRASIL               PR   
71218                0                B          NaN               PR   
71219                0                B          NaN               PR   
71220                0                B          NaN               PR   

         NM_CIDADE_NASC SG_UF_FERIAS NM_CIDADE_FERIAS  F_CD_ORIGEM_ALUNO  
1            MANDAGUARI          NaN              NaN                  0  
51              MARINGA           PR          MARINGA                  1  
809            LONDRINA           PR         MARIALVA                  0  
1489            MARINGA           --    NAO INFORMADO                  1  
1490            MARINGA           --    NAO INFORMADO                  1  
...                 ...          ...              ...                ...  
71216        BRAGANTINA          NaN          MARINGA                  0  
71217          CURITIBA          NaN        APUCARANA                  0  
71218       JACAREZINHO          NaN              NaN                  6  
71219          LONDRINA          NaN              NaN                  6  
71220  CANDIDO DE ABREU          NaN              NaN                  6  

[7322 rows x 8 columns]

Comecei a tratar, mas acho melhor voltar depois de tratar 'CD_ORIGEM_ALUNO', 'LT_NACIONALIDADE', 'NM_CIDADE_NASC', 'NM_PAIS_NASC', 'SG_UF_NASC_ALUNO'

In [77]:
colunas_a_remover.append('CD_ORIGEM_ALUNO')

### CD_TURNO

Mantém como está.

### DE_NATURALIDADE

Comentário do Alison:

> Acho que deveria ser a cidade de nascimento. Acho que este também não é um bom campo, pois também está em desuso, a maioria dos alunos estão com esse campo nulo. O que se usa mais é DE_NAT_ESTRANG o qual é a cidade de nascimento quando a pessoa é estrangeira e o próprio usuário digita como quiser, pois o campo NM_CIDADE_NASC é populado por um selectbox conforme cadastro dos correios e não tem cidades fora do Brasil.

Melhor desconsiderar esse campo.

In [78]:
colunas_a_remover.append('DE_NATURALIDADE')

### DH_MATRICULA

Mantém como está. Ano calendário

### DH_SITUACAO

Mantém como está. Ano calendário

Data da última situação do aluno

A partir de 2007 estará mais confiável. Antes pode ter informação errada. Foi feita uma tratativa para pegar o DH_MATRICULA se o DH_SITUACAO estivesse vazio.


Um caso: aluno entrou em 2014 e estava com a situação em branco, ou seja, matriculado. Quando um aluno tranca e volta, o DH_SITUACAO vai estar com a data da volta, e não da matrícula.

Outro caso: aluno com data de matrícula 2009 e data de ingresso 2006, o errado nesse caso seria o DH_MATRICULA. Antigamente era tudo manual.


### DT_ANO_INGRES

Mantém como está. Ano letivo

### DT_NASC_ALUNO

Vou manter apenas o ano (calendário)

In [79]:
df_final['DT_NASC_ALUNO'].value_counts(dropna=False)

DT_NASC_ALUNO
0001-01-01    62
1991-03-01    29
1990-10-10    28
1990-02-08    24
1994-06-22    23
              ..
1965-01-31     1
1969-03-31     1
1972-01-06     1
1979-12-18     1
1969-01-25     1
Name: count, Length: 11681, dtype: int64

In [80]:
df_final['F_DT_NASC_ALUNO'] = df_final['DT_NASC_ALUNO'].apply(lambda x: x.split('-')[0])

In [81]:
df_final[['DT_NASC_ALUNO', 'F_DT_NASC_ALUNO']]

DT_NASC_ALUNO F_DT_NASC_ALUNO
0        1974-07-08            1974
1        1971-02-08            1971
2        1971-07-12            1971
3        1974-05-25            1974
4        1966-07-27            1966
...             ...             ...
71216    1972-10-16            1972
71217    1981-01-16            1981
71218    2001-07-15            2001
71219    1999-12-13            1999
71220    2000-03-04            2000

[71221 rows x 2 columns]

In [82]:
colunas_a_remover.append('DT_NASC_ALUNO')

### LT_ESTADO_CIVIL

0 ou em branco = NÃO INFORMADO

1 = SOLTEIRO(A)

2 = CASADO(A)

3 = SEPARADO(A)

4 = DIVORCIADO(A)

5 = VIÚVO(A)

6 = OUTROS

In [83]:
df_final['LT_ESTADO_CIVIL'].value_counts(dropna=False).sort_index()

LT_ESTADO_CIVIL
0.0     1204
1.0    64671
2.0     3674
3.0     1001
4.0      278
5.0       15
6.0      146
NaN      232
Name: count, dtype: int64

In [84]:
df_final['LT_ESTADO_CIVIL'].fillna(0.0, inplace=True)

### LT_MODALIDADE

Como só tem um valor, desconsiderar coluna.

In [85]:
colunas_a_remover.append('LT_MODALIDADE')

### LT_NACIONALIDADE

Mantém como está? Vamos usar?

In [86]:
df_final['LT_NACIONALIDADE'].value_counts(dropna=False)

LT_NACIONALIDADE
B      70395
NaN      297
R        216
N        179
E        134
Name: count, dtype: int64

B= BRASILEIRA 

E= ESTRANGEIRA

N= BRASILEIRO NATURALIZADO

R= BRASILEIRO NASCIDO NO EXTERIOR

Nulo ou em branco = NÃO INFORMADO

### LT_RACA

Mantém como está

In [87]:
df_final['LT_RACA'].value_counts(dropna=False).sort_index()

LT_RACA
0    20818
1     3171
2    37476
3      177
4     1526
5     8053
Name: count, dtype: int64

0 = NÃO DECLARADA

1 = AMARELO(A), DE ORIGEM ORIENTAL

2 = BRANCO(A)

3 = INDÍGENA OU DE ORIGEM INDÍGENA

4 = NEGRO(A)

5 = PARDO(A)/MUDALTO(A)

Nulo ou em branco = NÃO INFORMADO

### NM_CIDADE_AULA

Desconsiderar. Muitas cidades, deveria ser só as cidades das regiões de cada campus.

In [88]:
colunas_a_remover.append('NM_CIDADE_AULA')

### NM_CIDADE_EST2GRAU, NM_CIDADE_FERIAS

Vamos usar?

Verificar se tem influência com a evasão antes de decidir

Separar campus sede dos outros

### NM_CIDADE_NASC

Vamos usar? Muitos nulos, como proceder?

In [89]:
df_final['NM_CIDADE_NASC'].value_counts(dropna=False)

NM_CIDADE_NASC
MARINGA           17790
NaN               13264
CIANORTE           1789
SAO PAULO          1703
UMUARAMA           1455
                  ...  
LIDIANOPOLIS          1
FARROUPILHA           1
PRES. EPITACIO        1
LIMA                  1
BRAGANTINA            1
Name: count, Length: 1502, dtype: int64

In [90]:
df_final[df_final['NM_PAIS_NASC'] != 'BRASIL']['NM_CIDADE_NASC'].value_counts(dropna=False)

NM_CIDADE_NASC
NaN               1274
MARINGA            198
LONDRINA            26
CIANORTE            25
SAO PAULO           24
                  ... 
JACAREI              1
GUATAMBU             1
URANIA               1
JOAQUIM TAVORA       1
JACAREZINHO          1
Name: count, Length: 401, dtype: int64

### NM_PAIS_NASC

In [91]:
df_final['NM_PAIS_NASC'].value_counts(dropna=False)

NM_PAIS_NASC
BRASIL             68872
NaN                 1751
JAPAO                342
ESTADOS UNIDOS        28
PARAGUAI              25
PORTUGAL              20
GUINE-BISSAU          17
CHINA                 15
ESPANHA               12
ARGENTINA             11
BOLIVIA               11
REPUBLICA DO HA       10
ALEMANHA               9
CHILE                  8
AFEGANISTAO            8
ITALIA                 7
PERU                   7
VENEZUELA              7
ANGOLA                 6
FRANCA                 6
CABO VERDE             5
COLOMBIA               4
TAIWAN                 3
CANADA                 2
BR                     2
INGLATERRA             2
HAITI                  2
S.TOME PRINCIPE        2
GUINE EQUATORIA        2
REP. DO HAITI          2
GANA                   1
GUINE BISSAU           1
BRASILEIRO             1
10\/11\/1999           1
REPUBLICA ARABE        1
COSTA DO MARFIM        1
BRAZIL                 1
27032001               1
FOZ DO IGUACU          1
MOCAMBIQUE  

In [92]:
NM_PAIS_NASC_nan = df_final[df_final['NM_PAIS_NASC'].isna()][colunas_loc_nasc_aluno]
NM_PAIS_NASC_nan['LT_NACIONALIDADE'].value_counts(dropna=False)

LT_NACIONALIDADE
B      1731
NaN      13
E         6
N         1
Name: count, dtype: int64

Preenchendo essa coluna com base na coluna LT_NACIONALIDADE

In [93]:
df_final['F_NM_PAIS_NASC'] = df_final.apply(lambda x: 'BRASIL' if x['LT_NACIONALIDADE'] == 'B' else x['NM_PAIS_NASC'], axis=1)

In [94]:
df_final['F_NM_PAIS_NASC'].value_counts(dropna=False)

F_NM_PAIS_NASC
BRASIL             70717
JAPAO                262
ESTADOS UNIDOS        26
PARAGUAI              22
NaN                   20
GUINE-BISSAU          17
CHINA                 15
PORTUGAL              15
BOLIVIA               11
REPUBLICA DO HA       10
ARGENTINA              9
ESPANHA                9
ALEMANHA               9
VENEZUELA              7
CHILE                  7
PERU                   7
FRANCA                 6
ANGOLA                 6
ITALIA                 6
CABO VERDE             5
COLOMBIA               4
TAIWAN                 3
INGLATERRA             2
HAITI                  2
GUINE EQUATORIA        2
REP. DO HAITI          2
CANADA                 2
S.TOME PRINCIPE        2
MOCAMBIQUE             1
GANA                   1
AUSTRIA                1
REINO UNIDO            1
COSTA DO MARFIM        1
REPUBLICA ARABE        1
GUINE BISSAU           1
JORDANIA               1
BENIN                  1
EUA                    1
GUATEMALA              1
REP.DEM.CO

In [95]:
df_final[df_final['F_NM_PAIS_NASC'].isna()][colunas_loc_nasc_aluno + ['F_NM_PAIS_NASC']]

CD_ORIGEM_ALUNO LT_NACIONALIDADE NM_PAIS_NASC SG_UF_NASC_ALUNO  \
128                  6                E          NaN              NaN   
129                  6                E          NaN              NaN   
17112                6                E          NaN              NaN   
17115                6                E          NaN              NaN   
17116                6                E          NaN              NaN   
32721                0                N          NaN               SP   
51119                0              NaN          NaN               SP   
51849                0              NaN          NaN               PR   
52625                0              NaN          NaN               PR   
54252                0              NaN          NaN               PR   
56077                0              NaN          NaN               MT   
60258                0              NaN          NaN               PR   
60767                0              NaN          NaN               SP   
61184                0              NaN          NaN               PR   
61504                0              NaN          NaN               PR   
62047                0              NaN          NaN               PR   
62285                0              NaN          NaN               PR   
62417                0              NaN          NaN               PR   
62493                0              NaN          NaN               PR   
62690                1                E          NaN              NaN   

      NM_CIDADE_NASC SG_UF_FERIAS   NM_CIDADE_FERIAS F_NM_PAIS_NASC  
128           BISSAU           PR            MARINGA            NaN  
129           BISSAU           PR            MARINGA            NaN  
17112   GUINE-BISSAU           PR            MARINGA            NaN  
17115            NaN           PR            MARINGA            NaN  
17116   GUINE-BISSAU           PR            MARINGA            NaN  
32721      SAO PAULO           SP              BAURU            NaN  
51119            NaN           SP         ADAMANTINA            NaN  
51849            NaN           PR          DOURADINA            NaN  
52625            NaN           PR            MARINGA            NaN  
54252            NaN           PR          APUCARANA            NaN  
56077            NaN           MT             CUIABA            NaN  
60258            NaN           PR          TERRA BOA            NaN  
60767            NaN           SP            MARILIA            NaN  
61184            NaN           PR           ROLANDIA            NaN  
61504            NaN           PR          APUCARANA            NaN  
62047            NaN           PR           COLORADO            NaN  
62285            NaN           PR           LONDRINA            NaN  
62417            NaN           PR  CRUZEIRO DO OESTE            NaN  
62493            NaN          NaN            MARINGA            NaN  
62690            NaN          NaN                NaN            NaN

Ainda restam alguns nulos que podemos preencher com a coluna SG_UF_NASC_ALUNO

In [96]:
df_final['F_NM_PAIS_NASC'] = df_final.apply(lambda x: 'BRASIL' if (not pd.isna(x['SG_UF_NASC_ALUNO'])) and pd.isna(x['F_NM_PAIS_NASC'])  else x['F_NM_PAIS_NASC'], axis=1)

In [97]:
df_final[df_final['F_NM_PAIS_NASC'].isna()][colunas_loc_nasc_aluno + ['F_NM_PAIS_NASC']]

CD_ORIGEM_ALUNO LT_NACIONALIDADE NM_PAIS_NASC SG_UF_NASC_ALUNO  \
128                  6                E          NaN              NaN   
129                  6                E          NaN              NaN   
17112                6                E          NaN              NaN   
17115                6                E          NaN              NaN   
17116                6                E          NaN              NaN   
62690                1                E          NaN              NaN   

      NM_CIDADE_NASC SG_UF_FERIAS NM_CIDADE_FERIAS F_NM_PAIS_NASC  
128           BISSAU           PR          MARINGA            NaN  
129           BISSAU           PR          MARINGA            NaN  
17112   GUINE-BISSAU           PR          MARINGA            NaN  
17115            NaN           PR          MARINGA            NaN  
17116   GUINE-BISSAU           PR          MARINGA            NaN  
62690            NaN          NaN              NaN            NaN

Existem ainda NaN cuja NM_CIDADE_NASC está BISSAU ou GUINE-BISSAU ainda podemos substituir. Bissau é a capital de Guiné-Bissau, então vou substituir todos esses por GUINE-BISSAU.

In [98]:
df_final['F_NM_PAIS_NASC'] = df_final.apply(lambda x: 'GUINE-BISSAU' if (not pd.isna(x['NM_CIDADE_NASC'])) and pd.isna(x['F_NM_PAIS_NASC'])  else x['F_NM_PAIS_NASC'], axis=1)

In [99]:
df_final[df_final['F_NM_PAIS_NASC'].isna()][colunas_loc_nasc_aluno + ['F_NM_PAIS_NASC']]

CD_ORIGEM_ALUNO LT_NACIONALIDADE NM_PAIS_NASC SG_UF_NASC_ALUNO  \
17115                6                E          NaN              NaN   
62690                1                E          NaN              NaN   

      NM_CIDADE_NASC SG_UF_FERIAS NM_CIDADE_FERIAS F_NM_PAIS_NASC  
17115            NaN           PR          MARINGA            NaN  
62690            NaN          NaN              NaN            NaN

Esses dois NaN que sobraram não tem muito o que fazer...

In [100]:
colunas_a_remover.append('NM_PAIS_NASC')
colunas_loc_nasc_aluno.remove('NM_PAIS_NASC')
colunas_loc_nasc_aluno.append('F_NM_PAIS_NASC')

### NU_SEMESTRE2_VEST

Ver em conjunto com TP_INGRES_CURSO. Talvez usaremos apenas esse campo, que já é basicamente o TP_INGRES_CURSO agrupado.

É o processo seletivo que o candidato fez para se tornar aluno, e atualmente usamos esses valores:

I = Vestibular de Inverno

V = Vestibular de Verão

A = PAS

D = Vestibular EAD

R = Vagas Remanescentes para Candidatos do Vestibular

E = Vagas Remanescentes para quem fez ENEM

G = Vagas Remanescentes para Refugiados ou Imigrantes em Situação de Vulnerabilidade 

S = SiSU

In [101]:
df_final['NU_SEMESTRE2_VEST'].value_counts(dropna=False)

NU_SEMESTRE2_VEST
V      31836
I      27573
A       5781
NaN     4141
E       1128
S        373
R        360
G         22
2          7
Name: count, dtype: int64

In [102]:
df_final[['NU_SEMESTRE2_VEST', 'TP_INGRES_CURSO', 'ST_ALUNO']].head(20)

NU_SEMESTRE2_VEST  TP_INGRES_CURSO ST_ALUNO
0                  I                1        2
1                NaN                6        7
2                  V                1        7
3                NaN                6        8
4                NaN               11        9
5                  I               14        7
6                NaN                6        2
7                NaN                6        2
8                  V                1        2
9                  I                1        8
10                 I                1      NaN
11                 I                1        J
12                 I                1        8
13                 V                1        8
14                 I               10        F
15                 I               10        8
16                 V                1        8
17               NaN                6        7
18               NaN                6        J
19                 V               14      NaN

TP_INGRES_CURSO encontrados quando NU_SEMESTRE2_VEST é nulo.

In [103]:
df_final[df_final['NU_SEMESTRE2_VEST'].isna()]['TP_INGRES_CURSO'].value_counts(dropna=False)

TP_INGRES_CURSO
6     1243
32    1113
3      996
5      384
8      105
22      97
7       50
4       49
53      32
26      24
10      22
56       9
12       4
1        4
11       3
19       2
24       1
16       1
29       1
27       1
Name: count, dtype: int64

5 - PORTADOR DE DIPLOMA PARA OUTRO CURSO
6 - PORTADOR DE DIPLOMA PARA NOVA HABILITACAO
32 - NOVA HABILITAÇÃO - CONTINUIDADE

7 - CONVENIO CULTURAL

3 - TRANSFERENCIA EXTERNA
4 - TRANSFERENCIA EX-OFFICIO
8 - TRANSFERENCIA INTERNA DE CURSO
10 - TRANSFERENCIA DE CAMPUS PARA MESMO CURSO
11 - TRANSFERENCIA DE CAMPUS PARA OUTRO CURSO
26 - PERMUTA DE CAMPUS
53 - TRANSFERENCIA INTERNA INDIGENA
56 - TRANSFERENCIA EXTERNA INDIGENA

1 - VESTIBULAR
12 - PROCESSO SELETIVO
16 - VESTIBULAR - 2 CHAMADA EXTRAORDINARIA
19 - VESTIBULAR - 5 CHAMADA EXTRAORDINARIA
22 - CONCURSO VESTIBULAR - INDIGENA
24 - VESTIBULAR - 6 CHAMADA EXTRAORDINARIA
29 - VESTIBULAR - 3 CHAMADA

27 - MATRICULA SUB-JUDICE

In [104]:
portador_diploma = [5, 6, 32]
convenio_cultural = [7]
transferencia = [3, 4, 8, 10, 11, 26, 53, 56]
vestibular = [1, 12, 16, 19, 22, 24, 29]
subjudice = [27]

Verificando o que tem na coluna NU_SEMESTRE2_VEST para cada valor de TP_INGRES_CURSO que foi econtrado com nulo em NU_SEMESTRE2_VEST

In [105]:
df_final[(df_final['NU_SEMESTRE2_VEST'].isna()) & (df_final['TP_INGRES_CURSO'] == 12)]

ANO_OCORRENCIA    CD_ALUNO  CD_CURSO  CD_ORIGEM_ALUNO CD_TURNO  \
12297            2011  1746001157        12                1        M   
15734            2009  1002268555        13                4        M   
15737            2009  1153058270        13                3        M   
15745            2011  1555422179        12                3        M   

      DE_NATURALIDADE DH_MATRICULA DH_SITUACAO  DT_ANO_INGRES DT_NASC_ALUNO  \
12297      BRASILEIRO   2007-02-01  2011-07-19           2007    1983-08-28   
15734      BRASILEIRA   2007-02-01  2018-09-25           2007    1987-04-18   
15737      BRASILEIRA   2007-02-01  2009-07-21           2007    1972-10-06   
15745      BRASILEIRA   2007-02-01  2011-12-23           2007    1988-07-25   

       LT_ESTADO_CIVIL LT_MODALIDADE LT_NACIONALIDADE  LT_RACA  \
12297              1.0             P                B        0   
15734              1.0             P                B        0   
15737              2.0             P                B        0   
15745              1.0             P                B        2   

           NM_CIDADE_AULA NM_CIDADE_EST2GRAU    NM_CIDADE_FERIAS  \
12297             MARINGA            MARINGA             MARINGA   
15734  RIBEIRO DOS INDIOS                NaN  RIBEIRO DOS INDIOS   
15737            CIANORTE                NaN            CIANORTE   
15745   CRUZEIRO DO OESTE           UMUARAMA   CRUZEIRO DO OESTE   

          NM_CIDADE_NASC NM_PAIS_NASC NU_SEMESTRE2_VEST SG_SEXO_ALUNO  \
12297            MARINGA       BRASIL               NaN             M   
15734               LINS       BRASIL               NaN             F   
15737                NaN       BRASIL               NaN             F   
15745  CRUZEIRO DO OESTE       BRASIL               NaN             F   

      SG_UF_AULA SG_UF_EST2GRAU SG_UF_FERIAS SG_UF_NASC_ALUNO ST_ALUNO  \
12297         PR             PR           PR               PR        8   
15734         SP            NaN           SP               SP        6   
15737         AC            NaN           PR               PR        8   
15745         PR             PR           PR               PR        7   

      TP_DEFIC_INEP TP_ESCOLA  TP_INGRES_CURSO TP_OPCAO_COTAS  \
12297      SNNNNNNN        GI               12              N   
15734           NaN         P               12              N   
15737           NaN        GI               12              N   
15745      NNNNNNNN         P               12              N   

       F_CD_ORIGEM_ALUNO F_DT_NASC_ALUNO F_NM_PAIS_NASC  
12297                  1            1983         BRASIL  
15734                  4            1987         BRASIL  
15737                  0            1972         BRASIL  
15745                  0            1988         BRASIL

In [106]:
df_final[df_final['TP_INGRES_CURSO'].isin(portador_diploma)]['NU_SEMESTRE2_VEST'].value_counts(dropna=False)

NU_SEMESTRE2_VEST
NaN    2740
V       106
Name: count, dtype: int64

In [107]:
df_final[df_final['TP_INGRES_CURSO'].isin(convenio_cultural)]['NU_SEMESTRE2_VEST'].value_counts(dropna=False)

NU_SEMESTRE2_VEST
NaN    50
Name: count, dtype: int64

In [108]:
df_final[df_final['TP_INGRES_CURSO'].isin(transferencia)]['NU_SEMESTRE2_VEST'].value_counts(dropna=False)

NU_SEMESTRE2_VEST
NaN    1240
I       102
V       101
A         3
Name: count, dtype: int64

In [109]:
df_final[df_final['TP_INGRES_CURSO'].isin(vestibular)]['NU_SEMESTRE2_VEST'].value_counts(dropna=False)

NU_SEMESTRE2_VEST
V      25725
I      21177
A       4143
NaN      110
2          3
Name: count, dtype: int64

In [110]:
df_final[df_final['TP_INGRES_CURSO'].isin(subjudice)]['NU_SEMESTRE2_VEST'].value_counts(dropna=False)

NU_SEMESTRE2_VEST
I      2
V      1
NaN    1
Name: count, dtype: int64

### SG_SEXO_ALUNO

Mantém como está

### SG_UF_AULA, SG_UF_EST2GRAU

Vamos usar?

### SG_UF_FERIAS

In [111]:
df_final['SG_UF_FERIAS'].value_counts(dropna=False)

SG_UF_FERIAS
PR     56591
SP      9339
NaN     3443
MS       669
MT       295
MG       238
SC       170
GO        89
RO        88
AC        63
RS        46
RJ        42
BA        40
DF        21
PA        13
TO        12
ES        11
CE         7
MA         6
--         6
P          5
RN         5
PI         4
PB         4
PE         3
AM         3
RR         2
S          2
SE         1
R          1
AL         1
AP         1
Name: count, dtype: int64

In [112]:
df_final['SG_UF_FERIAS'].unique(), df_final['SG_UF_FERIAS'].nunique(dropna=False)

(array(['PR', nan, 'SP', 'MG', 'SC', 'MT', 'GO', 'AC', 'MS', 'RS', 'RO',
        '--', 'S', 'PA', 'P', 'BA', 'PB', 'RJ', 'ES', 'TO', 'DF', 'RN',
        'CE', 'SE', 'MA', 'PE', 'RR', 'R', 'AL', 'PI', 'AM', 'AP'],
       dtype=object),
 32)

In [113]:
estados_a_remover_uf_ferias = ['--', 'S', 'P', 'R']

df_final[df_final['SG_UF_FERIAS'].isin(estados_a_remover_uf_ferias)]['SG_UF_FERIAS'].value_counts(dropna=False)

SG_UF_FERIAS
--    6
P     5
S     2
R     1
Name: count, dtype: int64

In [114]:
df_final[df_final['SG_UF_FERIAS'].isin(estados_a_remover_uf_ferias)]['NM_CIDADE_FERIAS'].value_counts(dropna=False)

NM_CIDADE_FERIAS
NAO INFORMADO        6
S\U001AO PAULO       1
CAMPO MOUR\U001AO    1
JACAREZINHO          1
GUARACI              1
MARINGA              1
AQUIDABAN            1
HOLAMBRA             1
PALMAS               1
Name: count, dtype: int64

In [115]:
df_final[df_final['SG_UF_FERIAS'].isin(estados_a_remover_uf_ferias)][colunas_loc_nasc_aluno + ['NM_CIDADE_FERIAS', 'SG_UF_FERIAS']]

CD_ORIGEM_ALUNO LT_NACIONALIDADE SG_UF_NASC_ALUNO NM_CIDADE_NASC  \
1489                 0                B               PR        MARINGA   
1490                 0                B               PR        MARINGA   
1491                 0                B               PR        MARINGA   
1857                 4                B               SP            NaN   
2003                 3                B               PR            NaN   
2189                 3                B               PR            NaN   
8566                 3                B               PR            NaN   
9086                 1                B               PR            NaN   
11382                1                B               PR            NaN   
16689                4                B               SP            NaN   
18665                0                B               SP          BAURU   
22683                0                B               PR        GOIOERE   
23879                3                B               PR         PALMAS   
69828                0                B               CE      FORTALEZA   

      SG_UF_FERIAS   NM_CIDADE_FERIAS F_NM_PAIS_NASC   NM_CIDADE_FERIAS  \
1489            --      NAO INFORMADO         BRASIL      NAO INFORMADO   
1490            --      NAO INFORMADO         BRASIL      NAO INFORMADO   
1491            --      NAO INFORMADO         BRASIL      NAO INFORMADO   
1857             S     S\U001AO PAULO         BRASIL     S\U001AO PAULO   
2003             P  CAMPO MOUR\U001AO         BRASIL  CAMPO MOUR\U001AO   
2189             P        JACAREZINHO         BRASIL        JACAREZINHO   
8566             P            GUARACI         BRASIL            GUARACI   
9086             P            MARINGA         BRASIL            MARINGA   
11382            P          AQUIDABAN         BRASIL          AQUIDABAN   
16689            S           HOLAMBRA         BRASIL           HOLAMBRA   
18665           --      NAO INFORMADO         BRASIL      NAO INFORMADO   
22683           --      NAO INFORMADO         BRASIL      NAO INFORMADO   
23879            R             PALMAS         BRASIL             PALMAS   
69828           --      NAO INFORMADO         BRASIL      NAO INFORMADO   

      SG_UF_FERIAS  
1489            --  
1490            --  
1491            --  
1857             S  
2003             P  
2189             P  
8566             P  
9086             P  
11382            P  
16689            S  
18665           --  
22683           --  
23879            R  
69828           --

In [116]:
def set_SG_UF_FERIAS_ALUNO_1(linha):
    if linha['NM_CIDADE_NASC'] in ['MARINGA', 'CAMPO MOURAO', 'JACAREZINHO', 'PALMAS', 'AQUIDABAN', 'GUARACI']:
        return 'PR'
    elif linha['NM_CIDADE_NASC'] in ['SAO PAULO', 'HOLAMBRA', 'SAO PAULO']:
        return 'SP'
    elif not pd.isna(linha['SG_UF_FERIAS']):
        return linha['SG_UF_FERIAS']
    else:
        return np.nan

In [117]:
df_final['F_SG_UF_FERIAS'] = df_final.apply(set_SG_UF_FERIAS_ALUNO_1, axis=1)

In [118]:
df_final['F_SG_UF_FERIAS'] = df_final['F_SG_UF_FERIAS'].replace({'P':'PR', 'R':'PR', 'S':'SP'})

In [119]:
df_final[df_final['F_SG_UF_FERIAS'].isna()].shape[0]

3272

Preenchidos os estados com uma letra só, colocando a sigla completa

In [120]:
df_final[df_final['NM_CIDADE_FERIAS'] == 'NAO INFORMADO'][colunas_loc_nasc_aluno + ['NM_CIDADE_FERIAS', 'SG_UF_FERIAS', 'F_SG_UF_FERIAS']]

CD_ORIGEM_ALUNO LT_NACIONALIDADE SG_UF_NASC_ALUNO NM_CIDADE_NASC  \
1489                 0                B               PR        MARINGA   
1490                 0                B               PR        MARINGA   
1491                 0                B               PR        MARINGA   
18665                0                B               SP          BAURU   
22683                0                B               PR        GOIOERE   
69828                0                B               CE      FORTALEZA   

      SG_UF_FERIAS NM_CIDADE_FERIAS F_NM_PAIS_NASC NM_CIDADE_FERIAS  \
1489            --    NAO INFORMADO         BRASIL    NAO INFORMADO   
1490            --    NAO INFORMADO         BRASIL    NAO INFORMADO   
1491            --    NAO INFORMADO         BRASIL    NAO INFORMADO   
18665           --    NAO INFORMADO         BRASIL    NAO INFORMADO   
22683           --    NAO INFORMADO         BRASIL    NAO INFORMADO   
69828           --    NAO INFORMADO         BRASIL    NAO INFORMADO   

      SG_UF_FERIAS F_SG_UF_FERIAS  
1489            --             PR  
1490            --             PR  
1491            --             PR  
18665           --             --  
22683           --             --  
69828           --             --

In [121]:
sg_ferias_nulo_cidade = df_final[(pd.isna(df_final['F_SG_UF_FERIAS'])) & (~pd.isna(df_final['NM_CIDADE_FERIAS']))][colunas_loc_nasc_aluno + [ 'SG_UF_FERIAS', 'F_SG_UF_FERIAS']]
sg_ferias_nulo_cidade

CD_ORIGEM_ALUNO LT_NACIONALIDADE SG_UF_NASC_ALUNO NM_CIDADE_NASC  \
106                  1                B               SP      GUARULHOS   
332                  4                B               SP            NaN   
706                  1                B               PR            NaN   
934                  2                B               PR            NaN   
1182                 3                B               PR            NaN   
...                ...              ...              ...            ...   
71203                0                B               RN          MACAU   
71205                0                B               PR       LONDRINA   
71208                0                B               RO    PORTO VELHO   
71216                0                B               PR     BRAGANTINA   
71217                0                B               PR       CURITIBA   

      SG_UF_FERIAS NM_CIDADE_FERIAS F_NM_PAIS_NASC SG_UF_FERIAS F_SG_UF_FERIAS  
106            NaN          MARINGA         BRASIL          NaN            NaN  
332            NaN      VOTUPORANGA         BRASIL          NaN            NaN  
706            NaN          MARINGA         BRASIL          NaN            NaN  
934            NaN         SANTA FE         BRASIL          NaN            NaN  
1182           NaN       PARANACITY         BRASIL          NaN            NaN  
...            ...              ...            ...          ...            ...  
71203          NaN          MARINGA         BRASIL          NaN            NaN  
71205          NaN         LONDRINA         BRASIL          NaN            NaN  
71208          NaN          MARINGA         BRASIL          NaN            NaN  
71216          NaN          MARINGA         BRASIL          NaN            NaN  
71217          NaN        APUCARANA         BRASIL          NaN            NaN  

[414 rows x 9 columns]

In [122]:
sg_ferias_nulo_cidade['NM_CIDADE_FERIAS'].value_counts(dropna=False)

NM_CIDADE_FERIAS
MARINGA              118
UMUARAMA              20
SARANDI                9
MANDAGUARI             8
APUCARANA              7
                    ... 
FARTURA                1
QUARTO CENTENARIO      1
HERCULANDIA            1
SERTAOZINHO            1
BOM SUCESSO            1
Name: count, Length: 192, dtype: int64

In [124]:
def set_SG_UF_FERIAS_2(linha, df_estados_municipios=df_estados_municipios):
    if pd.isna(linha['F_SG_UF_FERIAS']):
        try:
            return df_estados_municipios[df_estados_municipios['nome'] == linha['NM_CIDADE_FERIAS']]['uf'].values[0]

        except:
            return linha['F_SG_UF_FERIAS']
    else: 
        return linha['F_SG_UF_FERIAS']

In [125]:
df_final['F_SG_UF_FERIAS'] = df_final.apply(set_SG_UF_FERIAS_2, axis=1)

In [126]:
df_final[df_final['F_SG_UF_FERIAS'].isna()].shape[0]

2909

Após usar datasets de estados e cidades brasileiras, ainda ficamos com 2909 nulos.

In [127]:
colunas_a_remover.append('SG_UF_FERIAS')
colunas_loc_nasc_aluno.remove('SG_UF_FERIAS')
colunas_loc_nasc_aluno.append('F_SG_UF_FERIAS')

### SG_UF_NASC_ALUNO

In [128]:
df_final['SG_UF_NASC_ALUNO'].value_counts(dropna=False)

SG_UF_NASC_ALUNO
PR     51503
SP     13942
MS      1071
NaN      962
MT       656
MG       634
SC       453
RO       365
RS       332
RJ       324
GO       187
BA       138
PA       126
DF        88
PE        74
CE        65
AM        49
PB        40
ES        37
MA        31
AL        25
RN        25
TO        23
AP        17
PI        15
AC        15
SE        14
RR         6
WP         1
1          1
PY         1
5          1
Name: count, dtype: int64

In [129]:
df_final['SG_UF_NASC_ALUNO'].unique(), len(df_final['SG_UF_NASC_ALUNO'].unique())

(array(['PR', 'SP', 'RS', 'PB', 'MS', 'RJ', 'SC', 'MG', nan, 'MT', 'GO',
        'RO', 'DF', 'PA', 'PE', 'MA', 'BA', 'AL', 'AM', 'PI', 'ES', 'AC',
        '5', 'CE', 'SE', 'AP', 'RN', '1', 'WP', 'TO', 'PY', 'RR'],
       dtype=object),
 32)

Temos alguns valores que não são siglas de estados, logo devem ser tratados

In [130]:
estados_a_remover = ['5', '1', 'WP', 'PY', np.nan]

df_estados_a_remover = df_final[df_final['SG_UF_NASC_ALUNO'].isin(estados_a_remover)][colunas_loc_nasc_aluno]

df_estados_a_remover[df_estados_a_remover['SG_UF_NASC_ALUNO'] == '5']	

CD_ORIGEM_ALUNO LT_NACIONALIDADE SG_UF_NASC_ALUNO NM_CIDADE_NASC  \
4074                5                B                5            NaN   

     NM_CIDADE_FERIAS F_NM_PAIS_NASC F_SG_UF_FERIAS  
4074          COLIDER         BRASIL             MT

In [131]:
df_estados_a_remover[df_estados_a_remover['F_NM_PAIS_NASC'] == 'BRASIL']

CD_ORIGEM_ALUNO LT_NACIONALIDADE SG_UF_NASC_ALUNO NM_CIDADE_NASC  \
133                  6                B              NaN            NaN   
1014                 5                B              NaN            NaN   
2596                 3                B              NaN            NaN   
3242                 3                B              NaN            NaN   
3328                 0                B              NaN            NaN   
...                ...              ...              ...            ...   
71081                0                B              NaN            NaN   
71082                0                B              NaN            NaN   
71092                0                B              NaN            NaN   
71094                0                B              NaN            NaN   
71174                0                B              NaN            NaN   

      NM_CIDADE_FERIAS F_NM_PAIS_NASC F_SG_UF_FERIAS  
133            MARINGA         BRASIL             PR  
1014           MARINGA         BRASIL             PR  
2596    PRADO FERREIRA         BRASIL             PR  
3242               NaN         BRASIL            NaN  
3328               NaN         BRASIL            NaN  
...                ...            ...            ...  
71081          MARINGA         BRASIL             PR  
71082          MARINGA         BRASIL             PR  
71092          MARINGA         BRASIL             PR  
71094   NOVA ESPERANCA         BRASIL             PR  
71174         UMUARAMA         BRASIL             PR  

[532 rows x 7 columns]

In [132]:
df_estados_a_remover[df_estados_a_remover['F_NM_PAIS_NASC'] == 'BRASIL']['SG_UF_NASC_ALUNO'].value_counts(dropna=False)

SG_UF_NASC_ALUNO
NaN    528
5        1
1        1
WP       1
PY       1
Name: count, dtype: int64

531 alunos nascidos no Brasil tem o campo SG_UF_NASC_ALUNO nulo ou com valores inválidos. Todos os valores inválidos não nulos estão com os brasileiros.

In [133]:
df_estados_a_remover[df_estados_a_remover['F_NM_PAIS_NASC'] == 'BRASIL']['NM_CIDADE_NASC'].value_counts(dropna=False)

NM_CIDADE_NASC
NaN                    503
MARINGA                  7
ASSIS CHATEAUBRIAND      2
SAO PAULO                2
ALMADA                   2
IBARAKI                  1
ISESAKI                  1
YUKI-SHI                 1
SUZUKA                   1
SANTIAGO                 1
FUJIYOSHIDA              1
HARUE                    1
ALTO PARANA              1
CIANORTE                 1
FOZ DO IGUACU            1
BILAC                    1
GUAIRA                   1
TOLEDO                   1
CUIABA                   1
NANCY - FRANCA           1
OBU-SHI                  1
Name: count, dtype: int64

In [134]:
def set_SG_UF_NASC_ALUNO_1(linha):
    if linha['NM_CIDADE_NASC'] in ['MARINGA', 'TOLEDO', 'ASSIS CHATEAUBRIAND', 'ALTO PARANA', 'CIANORTE', 'FOZ DO IGUACU', 'GUAIRA', ]:
        return 'PR'
    elif linha['NM_CIDADE_NASC'] in ['SAO PAULO', 'BILAC']:
        return 'SP'
    elif linha['NM_CIDADE_NASC'] in ['CUIABA']:
        return 'MT'
    elif not pd.isna(linha['SG_UF_NASC_ALUNO']):
        return linha['SG_UF_NASC_ALUNO']
    else:
        return np.nan

In [135]:
df_final['F_SG_UF_NASC_ALUNO'] = df_final.apply(set_SG_UF_NASC_ALUNO_1, axis=1)

In [136]:
df_final['F_SG_UF_NASC_ALUNO'].value_counts(dropna=False)

F_SG_UF_NASC_ALUNO
PR     51547
SP     13934
MS      1069
NaN      935
MT       658
MG       632
SC       453
RO       362
RS       331
RJ       324
GO       187
BA       138
PA       126
DF        87
PE        74
CE        65
AM        47
PB        40
ES        37
MA        31
AL        25
RN        25
TO        23
AP        17
PI        15
AC        15
SE        14
RR         6
WP         1
1          1
PY         1
5          1
Name: count, dtype: int64

In [137]:
def set_SG_UF_NASC_ALUNO_2(linha, df_estados_municipios=df_estados_municipios, estados_a_remover=estados_a_remover):
    if pd.isna(linha['F_SG_UF_NASC_ALUNO']):
        try:
            return df_estados_municipios[df_estados_municipios['nome'] == linha['NM_CIDADE_NASC']]['uf'].values[0]

        except:
            return linha['F_SG_UF_NASC_ALUNO']
    else: 
        return linha['F_SG_UF_NASC_ALUNO']

In [138]:
df_final['F_SG_UF_NASC_ALUNO'] = df_final.apply(set_SG_UF_NASC_ALUNO_2, axis=1)

In [139]:
df_final[df_final['F_SG_UF_NASC_ALUNO'].isna()].shape[0]

921

In [140]:
colunas_a_remover.append('SG_UF_NASC_ALUNO')
colunas_loc_nasc_aluno.remove('SG_UF_NASC_ALUNO')
colunas_loc_nasc_aluno.append('F_SG_UF_NASC_ALUNO')

Essas colunas de localização ainda posso refinar mais...

### ST_ALUNO

NaN - Matriculado(a) 

0 - **EVASÃO** - Matrícula Cancelada pela UEM 

1 - Trancado(a) 

2 - **EVASÃO** - Matrícula Cancelada (Pelo aluno)

4 - **EVASÃO** - Transferido de Curso (Outro campus)

6 - **EVASÃO** - Transferido de IES

7 - Formado(a)

8 - **EVASÃO** - Cancelado por Abandono

9 - Jubilado(a) 

B - Transferido de Campus

F - Permuta Campus (Mesmo curso)

H - **EVASÃO** - Permuta Campus (Outro curso)

I - Falecido(a) 

J - **EVASÃO** - Matrícula Cancelada (Reprovação por falta por dois anos consecutivos)

K - **EVASÃO** - Matrícula Cancelada (Reprovação por nota duas vezes consecutivas – Regimento antigo)

M - Em Mobilidade Estudantil

O - **EVASÃO** - ND - Situação lançada em meados de 2001 para alunos com ocorrência de “Transferência interna de curso”. Pode incluir outras situações

P - **EVASÃO** - Transferido de Curso (Mesmo campus)



In [141]:
st_evasao = ['0', '2', '4', '6', '8', 'P', 'O', 'K', 'J', 'H']
st_nao_evasao = [np.nan, '1', '9', 'B', 'F', 'M', '7']

st_cancelamento = ['0', '2', '8', 'J', 'K']
st_transferencia = ['4', '6', 'H', 'O', 'P']
st_matriculado = [np.nan, '1', 'B', 'F' , 'M']
st_formado = ['7']
st_jubilado = ['9']
st_falecido = ['I']

In [142]:
df_final[df_final['ST_ALUNO'].isin(st_evasao) == True].shape[0]

26932

In [143]:
def set_BOOL_TRANCADO(linha):
    if linha['ST_ALUNO'] == '1':
        return 1
    return 0

In [144]:
df_final['BOOL_TRANCADO'] = df_final.apply(set_BOOL_TRANCADO, axis=1)

In [145]:
def set_BOOL_ST_ALUNO(linha:pd.Series) -> int:
    if linha['ST_ALUNO'] in st_evasao:
        return 1
    elif linha['ST_ALUNO'] in st_nao_evasao:
        return 0

In [146]:
df_final['BOOL_ST_ALUNO'] = df_final.apply(set_BOOL_ST_ALUNO, axis=1)

In [147]:
def set_MULTI_ST_ALUNO(linha:pd.Series) -> str:
    if linha['ST_ALUNO'] in st_formado:
        return 'FORMADO'
    elif linha['ST_ALUNO'] in st_cancelamento:
        return 'CANCELAMENTO'
    elif linha['ST_ALUNO'] in st_matriculado:
        return 'MATRICULADO'
    elif linha['ST_ALUNO'] in st_jubilado:
        return 'JUBILADO'
    elif linha['ST_ALUNO'] in st_transferencia:
        return 'TRANSFERENCIA'
    elif linha['ST_ALUNO'] in st_falecido:
        return 'FALECIDO'
    

In [148]:
df_final['MULTI_ST_ALUNO'] = df_final.apply(set_MULTI_ST_ALUNO, axis=1)

In [149]:
df_final['BOOL_ST_ALUNO'].value_counts(dropna=False)

BOOL_ST_ALUNO
0.0    44241
1.0    26932
NaN       48
Name: count, dtype: int64

Esses NaN são os falecidos.

In [150]:
df_final['BOOL_ST_ALUNO'].value_counts(normalize=True) * 100

BOOL_ST_ALUNO
0.0    62.159808
1.0    37.840192
Name: proportion, dtype: float64

In [151]:
pct_evadidos = df_final[df_final['BOOL_ST_ALUNO'] == 1].shape[0] / df_final.shape[0] * 100
pct_nao_evadidos = df_final[df_final['BOOL_ST_ALUNO'] == 0].shape[0] / df_final.shape[0] * 100

print(f'Percentual de evadidos: {pct_evadidos:.2f}%')
print(f'Percentual de não evadidos: {pct_nao_evadidos:.2f}%')

Percentual de evadidos: 37.81%
Percentual de não evadidos: 62.12%


In [152]:
df_final['MULTI_ST_ALUNO'].value_counts(dropna=False)

MULTI_ST_ALUNO
FORMADO          34016
CANCELAMENTO     26086
MATRICULADO       8981
JUBILADO          1244
TRANSFERENCIA      846
FALECIDO            48
Name: count, dtype: int64

In [153]:
df_final['MULTI_ST_ALUNO'].value_counts(normalize=True) * 100

MULTI_ST_ALUNO
FORMADO          47.761194
CANCELAMENTO     36.626838
MATRICULADO      12.610045
JUBILADO          1.746676
TRANSFERENCIA     1.187852
FALECIDO          0.067396
Name: proportion, dtype: float64

In [154]:
pct_formados = df_final[df_final['MULTI_ST_ALUNO'] == 'FORMADO'].shape[0] / df_final.shape[0] * 100
pct_cancelados = df_final[df_final['MULTI_ST_ALUNO'] == 'CANCELAMENTO'].shape[0] / df_final.shape[0] * 100
pct_matriculados = df_final[df_final['MULTI_ST_ALUNO'] == 'MATRICULADO'].shape[0] / df_final.shape[0] * 100
pct_jubilados = df_final[df_final['MULTI_ST_ALUNO'] == 'JUBILADO'].shape[0] / df_final.shape[0] * 100
pct_transferidos = df_final[df_final['MULTI_ST_ALUNO'] == 'TRANSFERENCIA'].shape[0] / df_final.shape[0] * 100
pct_falecidos = df_final[df_final['MULTI_ST_ALUNO'] == 'FALECIDO'].shape[0] / df_final.shape[0] * 100

print(f'Percentual de FORMADOS: {pct_formados:.2f}%')
print(f'Percentual de CANCELADOS: {pct_cancelados:.2f}%')
print(f'Percentual de MATRICULADOS: {pct_matriculados:.2f}%')
print(f'Percentual de JUBILADOS: {pct_jubilados:.2f}%')
print(f'Percentual de TRANSFERIDOS: {pct_transferidos:.2f}%')
print(f'Percentual de FALECIDOS: {pct_falecidos:.2f}%')

Percentual de FORMADOS: 47.76%
Percentual de CANCELADOS: 36.63%
Percentual de MATRICULADOS: 12.61%
Percentual de JUBILADOS: 1.75%
Percentual de TRANSFERIDOS: 1.19%
Percentual de FALECIDOS: 0.07%


Série temporal - evadidos no decorrer do tempo

### TP_DEFIC_INEP

In [155]:
df_final['TP_DEFIC_INEP'].unique()

array(['NNNNNNNN', nan, 'NNNSNNNN', 'NNNNNNNNNNNNN', 'SNNNNNNN',
       'SNSSNNNN', 'NNSNNNNN', 'SNSNNNNN', 'SNNNNNNNNNNNN', 'NNSSNNNN',
       'NNNNNSNN', 'NNNNNNNNSSNNN', 'SNNSNNNN', 'NNNNNNNNSNNNN',
       'SSNNNNNN', 'NSNNNNNN', 'NNSNNNSN', 'SNNNSNNN', 'NNNSNNNNNNNNN',
       'NNNNNNSN', 'NNSNNNNNNNNNN', 'NNNNSNNN', 'NNSNNNSNNNNNN',
       'SSNNNNNNNNNNN', 'NNNNNNNNSNNSN', 'NNNNSNNNNNNNN', 'NNNNNNNNNNNSN',
       'NNNNNNNNNSNNN', 'SNNNNNNNNNNSN', 'NNNNNNNNSSNSN', 'NSNNNNNNNNNNN',
       'SNNNSSNN'], dtype=object)

In [156]:
df_final['TP_DEFIC_INEP'].value_counts(dropna=False)

TP_DEFIC_INEP
NNNNNNNN         49921
NaN              16099
NNNNNNNNNNNNN     4027
SNNNNNNN           873
NNNSNNNN            77
NNSNNNNN            63
SNNNNNNNNNNNN       55
NNNNNNNNNNNSN       13
NNNSNNNNNNNNN       13
NNNNSNNN            12
SNSNNNNN             9
SNNSNNNN             7
NNNNNNNNNSNNN        7
NNSNNNNNNNNNN        6
NNNNNNNNSNNNN        6
NSNNNNNN             6
NNNNSNNNNNNNN        4
NNNNNNNNSSNNN        3
SSNNNNNN             3
NNSNNNSNNNNNN        2
NNSNNNSN             2
SNNNSNNN             2
NNSSNNNN             2
NNNNNNSN             1
SSNNNNNNNNNNN        1
NNNNNNNNSNNSN        1
NNNNNSNN             1
SNSSNNNN             1
SNNNNNNNNNNSN        1
NNNNNNNNSSNSN        1
NSNNNNNNNNNNN        1
SNNNSSNN             1
Name: count, dtype: int64

Como a maioria não tem deficiência, criar coluna boolrana para tem/não tem

In [210]:
df_final['TP_DEFIC_INEP'].fillna('NNNNNNNN').replace({r'[.*S+.*]' : True,
                                                      'NNNNNNNN' : False,
                                                      'NNNNNNNNNNNNN' : False}, regex=True).value_counts(dropna=False)

TP_DEFIC_INEP
False    70047
True      1174
Name: count, dtype: int64

Validado acima

In [214]:
df_final['BOOL_DEFIC_INEP'] = df_final['TP_DEFIC_INEP'].fillna('NNNNNNNN').replace({r'[.*S+.*]' : True,
                                                      'NNNNNNNN' : False,
                                                      'NNNNNNNNNNNNN' : False}, regex=True)

In [217]:
df_final['BOOL_DEFIC_INEP'].value_counts(dropna=False) 

BOOL_DEFIC_INEP
False    70047
True      1174
Name: count, dtype: int64

### TP_ESCOLA

In [157]:
df_final['TP_ESCOLA'].value_counts(dropna=False)

TP_ESCOLA
GI     33592
P      27342
E       4467
GP      3392
NaN     1154
SG       776
SP       266
F        107
EX        91
M         34
Name: count, dtype: int64

Os tipos F, E e M não são mais utilizados. Vou agregar todos em uma única categoria 'G'. Quanto aos nulos, não há o que fazer.

In [158]:
df_final['F_TP_ESCOLA'] = df_final['TP_ESCOLA'].replace({'E' : 'G', 'M' : 'G', 'F' : 'G'})

In [159]:
df_final['F_TP_ESCOLA'].value_counts(dropna=False)

F_TP_ESCOLA
GI     33592
P      27342
G       4608
GP      3392
NaN     1154
SG       776
SP       266
EX        91
Name: count, dtype: int64

In [160]:
colunas_a_remover.append('TP_ESCOLA')	

### TP_INGRES_CURSO

In [218]:
colunas_a_remover.append('TP_INGRES_CURSO')

Provavelmente não será utilizada da forma como está, muitos valores.

Agrupá-los de alguma forma. A coluna NU_SEMESTRE2_VEST já é essa agrupada.

Valores e grupos de TP_INGRESS_CURSO quando NU_SEMESTRE2_VEST.

5 - PORTADOR DE DIPLOMA PARA OUTRO CURSO
6 - PORTADOR DE DIPLOMA PARA NOVA HABILITACAO
32 - NOVA HABILITAÇÃO - CONTINUIDADE

7 - CONVENIO CULTURAL

3 - TRANSFERENCIA EXTERNA
4 - TRANSFERENCIA EX-OFFICIO
8 - TRANSFERENCIA INTERNA DE CURSO
10 - TRANSFERENCIA DE CAMPUS PARA MESMO CURSO
11 - TRANSFERENCIA DE CAMPUS PARA OUTRO CURSO
26 - PERMUTA DE CAMPUS
53 - TRANSFERENCIA INTERNA INDIGENA
56 - TRANSFERENCIA EXTERNA INDIGENA

1 - VESTIBULAR
12 - PROCESSO SELETIVO
16 - VESTIBULAR - 2 CHAMADA EXTRAORDINARIA
19 - VESTIBULAR - 5 CHAMADA EXTRAORDINARIA
22 - CONCURSO VESTIBULAR - INDIGENA
24 - VESTIBULAR - 6 CHAMADA EXTRAORDINARIA
29 - VESTIBULAR - 3 CHAMADA

27 - MATRICULA SUB-JUDICE

### TP_OPCAO_COTAS

In [161]:
df_final['TP_OPCAO_COTAS'].value_counts(dropna=False)

TP_OPCAO_COTAS
N      62808
S       5862
NaN     2277
O        127
G        126
F         21
Name: count, dtype: int64

Os nulos aqui são não informado. Vou considerar sem cotas, universal

In [162]:
df_final['F_TP_OPCAO_COTAS'] = df_final['TP_OPCAO_COTAS'].fillna('N')

In [163]:
colunas_a_remover.append('TP_OPCAO_COTAS')

Criando uma coluna booleana de cotas

In [164]:
df_final['BOOL_COTAS'] = df_final['F_TP_OPCAO_COTAS'].apply(lambda x: 0 if x == 'N' else 1)

In [165]:
df_final['BOOL_COTAS'].value_counts(dropna=False)

BOOL_COTAS
0    65085
1     6136
Name: count, dtype: int64

# Colunas

In [166]:
df_final.columns

Index(['ANO_OCORRENCIA', 'CD_ALUNO', 'CD_CURSO', 'CD_ORIGEM_ALUNO', 'CD_TURNO',
       'DE_NATURALIDADE', 'DH_MATRICULA', 'DH_SITUACAO', 'DT_ANO_INGRES',
       'DT_NASC_ALUNO', 'LT_ESTADO_CIVIL', 'LT_MODALIDADE', 'LT_NACIONALIDADE',
       'LT_RACA', 'NM_CIDADE_AULA', 'NM_CIDADE_EST2GRAU', 'NM_CIDADE_FERIAS',
       'NM_CIDADE_NASC', 'NM_PAIS_NASC', 'NU_SEMESTRE2_VEST', 'SG_SEXO_ALUNO',
       'SG_UF_AULA', 'SG_UF_EST2GRAU', 'SG_UF_FERIAS', 'SG_UF_NASC_ALUNO',
       'ST_ALUNO', 'TP_DEFIC_INEP', 'TP_ESCOLA', 'TP_INGRES_CURSO',
       'TP_OPCAO_COTAS', 'F_CD_ORIGEM_ALUNO', 'F_DT_NASC_ALUNO',
       'F_NM_PAIS_NASC', 'F_SG_UF_FERIAS', 'F_SG_UF_NASC_ALUNO',
       'BOOL_TRANCADO', 'BOOL_ST_ALUNO', 'MULTI_ST_ALUNO', 'F_TP_ESCOLA',
       'F_TP_OPCAO_COTAS', 'BOOL_COTAS'],
      dtype='object')

Idade no ingresso

Morar ou não na cidade do campus

Ser de outros estado



## MEDIDAS

### Idade na matrícula

In [167]:
from datetime import datetime, timedelta

In [168]:
df_final[df_final['DH_MATRICULA'].isna()]

Empty DataFrame
Columns: [ANO_OCORRENCIA, CD_ALUNO, CD_CURSO, CD_ORIGEM_ALUNO, CD_TURNO, DE_NATURALIDADE, DH_MATRICULA, DH_SITUACAO, DT_ANO_INGRES, DT_NASC_ALUNO, LT_ESTADO_CIVIL, LT_MODALIDADE, LT_NACIONALIDADE, LT_RACA, NM_CIDADE_AULA, NM_CIDADE_EST2GRAU, NM_CIDADE_FERIAS, NM_CIDADE_NASC, NM_PAIS_NASC, NU_SEMESTRE2_VEST, SG_SEXO_ALUNO, SG_UF_AULA, SG_UF_EST2GRAU, SG_UF_FERIAS, SG_UF_NASC_ALUNO, ST_ALUNO, TP_DEFIC_INEP, TP_ESCOLA, TP_INGRES_CURSO, TP_OPCAO_COTAS, F_CD_ORIGEM_ALUNO, F_DT_NASC_ALUNO, F_NM_PAIS_NASC, F_SG_UF_FERIAS, F_SG_UF_NASC_ALUNO, BOOL_TRANCADO, BOOL_ST_ALUNO, MULTI_ST_ALUNO, F_TP_ESCOLA, F_TP_OPCAO_COTAS, BOOL_COTAS]
Index: []

In [169]:
df_final[df_final['DT_NASC_ALUNO'].isna()]

Empty DataFrame
Columns: [ANO_OCORRENCIA, CD_ALUNO, CD_CURSO, CD_ORIGEM_ALUNO, CD_TURNO, DE_NATURALIDADE, DH_MATRICULA, DH_SITUACAO, DT_ANO_INGRES, DT_NASC_ALUNO, LT_ESTADO_CIVIL, LT_MODALIDADE, LT_NACIONALIDADE, LT_RACA, NM_CIDADE_AULA, NM_CIDADE_EST2GRAU, NM_CIDADE_FERIAS, NM_CIDADE_NASC, NM_PAIS_NASC, NU_SEMESTRE2_VEST, SG_SEXO_ALUNO, SG_UF_AULA, SG_UF_EST2GRAU, SG_UF_FERIAS, SG_UF_NASC_ALUNO, ST_ALUNO, TP_DEFIC_INEP, TP_ESCOLA, TP_INGRES_CURSO, TP_OPCAO_COTAS, F_CD_ORIGEM_ALUNO, F_DT_NASC_ALUNO, F_NM_PAIS_NASC, F_SG_UF_FERIAS, F_SG_UF_NASC_ALUNO, BOOL_TRANCADO, BOOL_ST_ALUNO, MULTI_ST_ALUNO, F_TP_ESCOLA, F_TP_OPCAO_COTAS, BOOL_COTAS]
Index: []

In [170]:
df_final['DT_NASC_ALUNO']

0        1974-07-08
1        1971-02-08
2        1971-07-12
3        1974-05-25
4        1966-07-27
            ...    
71216    1972-10-16
71217    1981-01-16
71218    2001-07-15
71219    1999-12-13
71220    2000-03-04
Name: DT_NASC_ALUNO, Length: 71221, dtype: object

In [206]:
diff = pd.to_datetime(df_final['DH_MATRICULA'], errors='coerce') - pd.to_datetime(df_final['DT_NASC_ALUNO'], errors='coerce')

def a(linha):
    if pd.isna(linha):
        return np.nan

    return np.floor(linha.days / 365)

df_final['IDADE_MATRICULA'] = diff.apply(a)

### Ano de nascimento

In [207]:
df_final['ANO_NASCIMENTO'] = df_final['DT_NASC_ALUNO'].apply(lambda x: datetime.strptime(x, '%Y-%m-%d').year)

### Estudou ou não no campus sede

In [221]:
df_cursos = pd.read_csv('arquivos/cursos.csv')
df_cursos.head()

CD_CURSO             NM_CURSO TP_HABILITACAO SG_CENTRO NM_CAMPUS  MAX_ANOS
0         1  Ciências Econômicas    BACHARELADO     CSA     MARINGA         8
1         2              Direito    BACHARELADO     CSA     MARINGA         8
2         3               Letras   LICENCIATURA     CCH     MARINGA         8
3         4            Geografia          AMBOS     CCH     MARINGA         8
4         5             História   LICENCIATURA     CCH     MARINGA         8

In [223]:
df_cursos['BOOL_SEDE'] = df_cursos.apply(lambda x: True if x['NM_CAMPUS'] == 'MARINGA' else False, axis=1)

### CD_ORIGEM_ALUNO

In [234]:
cidades_origem_maringa = df_final[df_final['CD_ORIGEM_ALUNO'] == 1][['NM_CIDADE_NASC', 'NM_CIDADE_FERIAS', 'NM_CIDADE_AULA', 'NM_CIDADE_EST2GRAU']]
cidades_origem_maringa

NM_CIDADE_NASC NM_CIDADE_FERIAS NM_CIDADE_AULA NM_CIDADE_EST2GRAU
0            MARINGA          MARINGA        MARINGA            MARINGA
2            MARINGA          MARINGA        MARINGA            MARINGA
6            MARINGA           ITAMBE        MARINGA                NaN
11           MARINGA          MARINGA        MARINGA            MARINGA
12           MARINGA          MARINGA        MARINGA            MARINGA
...              ...              ...            ...                ...
71111        MARINGA          MARINGA        MARINGA            SARANDI
71115        UBATUBA         CIANORTE       CIANORTE            UBATUBA
71139       UMUARAMA         UMUARAMA       UMUARAMA           UMUARAMA
71170       UMUARAMA         UMUARAMA       UMUARAMA       ALTO PIQUIRI
71179        MARINGA          MARINGA        MARINGA            MARINGA

[27404 rows x 4 columns]

A ideia era pegar esse campo (CD_ORIGEM_ALUNO) e verificar quais são as cidades que ele considera na região de Maringá para criar este campo para NM_CIDADE_NASC e NM_CIDADE_FERIAS. Porém, ao filtrarmos esse campo para Maringá, vemos entradas que em nenhum dos campos de cidade consta Maringá...

Ideias
* Passa as férias na cidade de seu campus? (Usar NM_CIDADE_FERIAS, já que NM_CIDADE_AULA não é útil)
* Aluno nasceu no PR, SP, Outros Estados ou Exterior (SG_UF_NASC_ALUNO)
* Da para fazer um igual ao de cima para SG_UF_FERIAS